RT: train cancelling

antoine-de opened the pull request:
handle realtime message to cancel a train (VehicleJourney)

To query on realtime data a new `journeys` parameter has been added:
`data_freshess` (the name can be discussed) that can take 3 values:
- base_schedule 
- adapted_schedule
- realtime

The `disruption_active` parameter is thus now deprecated

antoine-de committed:
Ed: create theoric VP

antoine-de committed:
RT: handle vj cancellation

TeXitoi commented on source/kraken/realtime.cpp:
```diff
@@ -29,11 +29,53 @@ www.navitia.io
+    auto circulation_date = boost::gregorian::from_undelimited_string(trip.start_date());
+
+    if (trip.CANCELED) {
+        cancel_vj(vj, circulation_date, trip_update, data);
```
I think you need to loop on every vj of the meta vj.

xlqian commented on source/jormungandr/jormungandr/interfaces/v1/Journeys.py:
```diff
@@ -624,6 +626,10 @@ def get(self, region=None, lon=None, lat=None, uri=None):
 
+        if args['data_freshness'] is None:
+            # retrocompatibilty handling
+            args['data_freshness'] = 'adapted' if args['disruption_active'] is True else 'theoric'
```
it should be `theoretical`.. let's keep it since it's used everywhere LOL

pbougue replied:
As it is the first time we expose it I vote for a change.
Other suggestion from Pascal : `revised` instead of adapted

antoine-de replied:
I personally vote for `scheduled` and `amended`, but I think it's better to wait for andrew

TeXitoi commented on source/routing/dataraptor.cpp:
```diff
@@ -90,29 +90,20 @@ void dataRAPTOR::load(const type::PT_Data &data)
+        for(const type::JourneyPattern* journey_pattern : data.journey_patterns) {
+            for(int i=0; i<=365; ++i) {
+                journey_pattern->for_each_vehicle_journey([&](const nt::VehicleJourney& vj) {
+                    if(vj.validity_patterns[rt_level]->check2(i)) {
```
spacingsss

TeXitoi commented on source/routing/raptor.cpp:
```diff
@@ -412,16 +412,8 @@ void RAPTOR::set_valid_jp_and_jpp(
-    case nt::RTLevel::RealTime:
-        throw navitia::exception("rt not implemented yet");
-    }
+
```
empty line

xlqian commented on source/jormungandr/jormungandr/scenarios/default.py:
```diff
@@ -100,10 +100,14 @@ def parse_journey_request(self, requested_type, request):
         req.journeys.wheelchair = request["wheelchair"]
-        if request["disruption_active"]:
+
+        if request["data_freshness"] == 'realtime':
```
single quote

xlqian commented on source/jormungandr/tests/kirin_realtime_tests.py:
```diff
@@ -45,21 +45,56 @@ def _make_mock_item(self, *args, **kwargs):
+        assert _get_used_vj(response) == [['vj:A:0'], []]
 
-        #TODO
+        self.send_mock("vj:A:0", "20120614", 'cancelled')
```
Minor: cancelled or canceled?

xlqian commented on source/jormungandr/tests/kirin_realtime_tests.py:
```diff
@@ -70,13 +105,13 @@ def make_mock_kirin_item(*args, **kwargs):
+    trip.start_date = date
+
+    if status == 'canceled':
+        trip.schedule_relationship = gtfs_realtime_pb2.TripDescriptor.CANCELED
```
Minor: cancelled or canceled?

antoine-de replied:
http://www.quickanddirtytips.com/education/grammar/canceled-or-cancelled :)

xlqian replied:
It was a question more about consistence than english grammar :)

antoine-de replied:
yes you're right, I choose the english version (canceled)

xlqian commented on source/kraken/realtime.cpp:
```diff
@@ -29,11 +29,53 @@ www.navitia.io
+
+    auto circulation_date = boost::gregorian::from_undelimited_string(trip.start_date());
+
+    if (trip.CANCELED) {
```
`trip.CANCELED` has been tested, no?

antoine-de replied:
yes, but the first one is temporary, since for the moment we handle only the CANCELED tag

xlqian replied:
ok

TeXitoi replied:
So you want a switch here...

xlqian commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -0,0 +1,186 @@
+
+namespace nt = navitia::type;
+
+transit_realtime::TripUpdate make_cancellation_message(std::string vj_uri, std::string date) {
```
`const std::string&`

antoine-de replied:
I don't think it's worth it since it is always used with a lvalue

TeXitoi replied:
It's not worst to use const&, and it's better in the general case.

You also need to put this method static.

xlqian commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -0,0 +1,186 @@
+    b.vj("A", "000001", "", true, "vj:1")("stop1", "08:00"_t)("stop2", "09:00"_t);
+
+    transit_realtime::TripUpdate trip_update = make_cancellation_message("vj:1", "20150928");
+    auto& pt_data = b.data->pt_data;
```
auto* pt_data

xlqian commented on source/routing/dataraptor.cpp:
```diff
@@ -90,29 +90,20 @@ void dataRAPTOR::load(const type::PT_Data &data)
+        const auto rt_level = level_cont.first;
+        auto& jp_vp = level_cont.second;
+        jp_vp.assign(366, boost::dynamic_bitset<>(data.journey_patterns.size()));
+        for(const type::JourneyPattern* journey_pattern : data.journey_patterns) {
```
const auto*? for consistence?

xlqian commented on source/ed/ed_reader.cpp:
```diff
@@ -822,6 +822,7 @@ void EdReader::fill_vehicle_journeys(nt::Data& data, pqxx::work& work){
         if(!const_it["validity_pattern_id"].is_null()){
             vj->validity_patterns[RTLevel::Theoric] =
                     validity_pattern_map[const_it["validity_pattern_id"].as<idx_t>()];
+            vj->validity_patterns[RTLevel::RealTime] = vj->validity_patterns[RTLevel::Theoric];
```
so, no real time information in db just now or there will never be?

antoine-de replied:
maybe there will, maybe there won't :)
why ?

pbougue replied:
At minimum, add a little comment for that.
But all this consistency-and-C++-model should be in a special builder (or in meta-vj), not in ED, builder, chaos and Kirin readers anymore

antoine-de replied:
I think this should be done in another PR (at least to wait for the JourneyPattern refactoring).

Link: http://jira.canaltp.fr/browse/NAVP-299

pbougue commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -0,0 +1,186 @@
+    auto res = compute(nt::RTLevel::Theoric);
+    BOOST_REQUIRE_EQUAL(res.size(), 1);
+
+    //on the realtime level, we should also gt one solution, since for the moment there is no cancellation
```
gt > `get`

antoine-de committed:
corrections after review

antoine-de committed:
All: rename Theoric to Base

theoric is not english and Base is better according to our great free
american consultant :wink:

antoine-de committed:
Type: rename metavj theoric vjs

antoine-de committed:
Jormungandr: change mock

when elastic search is run, the mock is called for end_point_id, so we
add it

TeXitoi commented:
OK for me.
